fix too many open files on Mac

zhili1208 opened the pull request:
the bug is when nuget restore a package which contains many dependencies on mac, will get too many open files error,

the problem here is reader.GetNuspec() read the nuspec stream from path, return it to NuspecReader but does not close the stream, we still hold the nuspec file.

zhili1208 committed:
fixed https://github.com/NuGet/Home/issues/1899

zhili1208 committed:
too many open files fix and test

emgarten committed:
Moving SimpleTestPackage helpers to test utility

emgarten committed:
P2P reference builder fixes and optimizations

emgarten committed:
Dependency target and type constraint improvements

emgarten committed:
External reference clean up and fixes

joelverhagen committed:
Add validation for source and support "All" (V2 compatibility)

emgarten committed:
Filter disabled sources from xplat

zhili1208 committed:
format

zhili1208 commented:
@emgarten @yishaigalatzer

emgarten commented:
:rocket:

emgarten commented on test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:
```diff
@@ -677,6 +677,40 @@ public class RestoreCommandTests
+        {
+            // Arrange
+            var sources = new List<PackageSource>();
+            sources.Add(new PackageSource("https://www.myget.org/F/aspnetcidev/api/v3/index.json"));
```
this feed changes a lot, we shouldn't depend on it here.

yishaigalatzer replied:
you can grab packages from there an put them on the volatile feed

emgarten replied:
Thinking about this more, it would probably be better to just use local packages. Once the simple package test utility is added from my PR it should be trivial to make a package with 500 dependencies and restore it.

emgarten commented:
@zhili1208 would you add a friendly error message if the nupkg and nuspec streams here are null? The nupkg can be null if there is a network issue, and potentially the package doesn't contain a nuspec. These are rare but I have seen null exceptions from this part of the code. it would be great to fix them as part of this also.

yishaigalatzer commented on src/NuGet.Core/NuGet.Protocol.Core.v3/RemoteRepositories/PackageUtilities.cs:
```diff
@@ -25,8 +25,9 @@ internal static class PackageUtilities
                     {
-                        return new NuspecReader(reader.GetNuspec());
+                        return new NuspecReader(nuspecStream);
                     }
```
Localize the string below while you are at it report.LogWarning($"The ZIP archive {fileStream.Name} is corrupt");

zhili1208 committed:
review update

deepakaravindr committed:
Functional tests revamp

Added a bunch of powershell scripts to do the following and more.
Parameterized the scripts as much as possible
such that nothing is hard-coded.

1) Pick up bits from the given branch, build number and so on
2) Install/Uninstall VSIX
3) Launch VS and start the functional tests

Implemented a bootstrap.ps1 that will used by CI to
copy and extract EndToEnd.zip, which contains all the necessary scripts to
run the functional tests

Added a copy tools and setup machine script to copy the necessary CI tools
and to set up registry keys specific to the CI machine. This is different
from the setup machine script intended to be run on the dev boxes

Install NuGet VSIX script does the following

1) Kills any running instances of VS
2) Uninstalls NuGet VSIX if installed
3) Installs the NuGet VSIX as indicated by $NuGetDropPath after copying
it to $FuncTestRoot

Launches VS and waits until DTE has become available. Controlled by
polling timeout and frequency

Runs the functional tests after obtaining DTE which is guaranteed
to be available. Also, provides real time monitoring. Finally, copies the
test results to the CI

analogrelay committed:
treat .exe files as valid assemblies

emgarten committed:
LibraryDependencyTarget clean up

emgarten committed:
External reference fixes

zhili1208 committed:
review change

zhili1208 committed:
update
